src/mgopurge/trim.go

```diff
@@ -0,0 +1,487 @@
+		}
+		if !foundAllDocs {
+			// We won't purge this transaction if one of the docs
+			// involved doesn't have a long transaction queues
+			tb.txnsSkippedCount++
+			delete(tb.txns, txnId)
+			continue
+		}
+		tb.txnsToRemove = append(tb.txnsToRemove, txnId)
+	}
+}
+
+// processDoc processes a single document to remove transactions that are present
+// in tb.txns. This can be called in parallel because it only touches the values
+// on this doc.
+func (tb *txnBatchTrimmer) processDoc(key docKey, doc *txnDoc) error {
+	tokensToPull := make([]interface{}, 0, len(tb.txnsToRemove))
+	tokensToSet := make([]string, 0, len(doc.queue))
+	remainingQueue := make([]parsedToken, 0, len(doc.queue))
+	txns := tb.txns
+	for _, tokenInfo := range doc.queue {
+		if _, ok := txns[tokenInfo.txnId]; ok {
+			// we ignore nonce, as we will pull all
+			// references to a given txn id, even if
+			// the nonce doesn't match
+			tokensToPull = append(tokensToPull, tokenInfo.token)
+		} else {
+			remainingQueue = append(remainingQueue, tokenInfo)
+			tokensToSet = append(tokensToSet, tokenInfo.token)
+		}
+	}
+	// for small numbers of tokens left in the queue it is faster to just
+	// set the absolute value. When trimming only a few it is slightly faster
+	// to pull only the ones that are being removed. Experimental tests show
+	// that the difference is small and the time is mostly dominated by the
+	// sheer number of items that are left.
+	// Ideally txnBatchSize is set high enough that you take a significant
+	// number of tokens in each pass.
+	// (extreme example is removing the last 100,000 in one pass, is nearly
+	// instant to set the value to [], but fairly slow to iterate all of
+	// them and pull them out.)
+	if len(tokensToPull)*10 > len(tokensToSet) {
+		// removing more than 10% of the tokens, we just set the whole value
+		// (empirically $set is much faster than $pullAll)
+		// note that $set isn't safe if doing concurrent accesses, but
+		// we don't allow that here. Controllers must be stopped, and
+		// only one goroutine modifies a doc at a time.
+		if err := tb.tokenSetter(key, tokensToSet, len(tokensToPull)); err != nil {
+			return err
+		}
+	} else {
+		if err := tb.tokenRemover(key, tokensToPull); err != nil {
+			return err
+		}
+	}
+	doc.queue = remainingQueue
+	return nil
+}
+
+// processDocs works through all of the docsToCleanup that were found in
+// checkTransactionsFindDocs, it removes all of the tokens that refer to transactions
+// that we are removing, and updates the in-memory cache so the doc objects no
+// longer refer to those tokens.
+func (tb *txnBatchTrimmer) processDocs() error {
+	if len(tb.docsToCleanup) == 0 {
+		return nil
+	}
+	errCh := make(chan error)
+	count := 0
+	// each doc is independent, so we fork it off into another goroutine
+	// and collect the results back onto the channel.
+	var err error
+	for key, doc := range tb.docsToCleanup {
+		count++
+		key := key
+		doc := doc
+		go func() {
```

Comment:
I'm concerned that this can start an unbounded number of goroutines, which might make things worse not better. It might be better to bound the amount of concurrency (utils/parallel.Run makes that quite easy)